File: scripts/tai-ap-14f2a/verify.py
from __future__ import annotations

import argparse
import json
import subprocess
from pathlib import Path
from typing import Any

ROOT = Path(__file__).resolve().parents[2]
BASE_COMMIT = "015d6fc4c1fdee24860a2ebee49d2cd95de9be2c"
BRANCH = "agent/tai-ap-14f2a-rosstat-fact-pack-3381"
ISSUE = 3381
PACK_ID = "factpack.rosstat.7708234640-vshp2016254"
SOURCE_ID = "official.rosstat.opendata.7708234640-vshp2016254"
EXPECTED_PATHS = sorted(
    [
        ".github/workflows/tai-ap-14f2a.yml",
        "apps/tai/tai/migrations/0022_public_fact_pack_authority.sql",
        "apps/tai/tai/migrations/manifest.json",
        "apps/tai/tai/rosstat_fact_pack.py",
        "apps/tai/tests/test_migration_manifest.py",
        "apps/tai/tests/test_rosstat_fact_pack.py",
        "apps/tai/tests/test_rosstat_fact_pack_database.py",
        "docs/platform-v7/autopilot/scopes/tai-ap-14f2a-3381.json",
        "scripts/tai-ap-14f2a/verify.py",
    ]
)


def fail(message: str) -> None:
    raise RuntimeError(f"TAI_AP_14F2A:{message}")


def read(path: str) -> str:
    return (ROOT / path).read_text(encoding="utf-8")


def read_json(path: str) -> dict[str, Any]:
    try:
        value = json.loads(read(path))
    except (OSError, json.JSONDecodeError) as error:
        fail(f"invalid JSON {path}: {error}")
    if not isinstance(value, dict):
        fail(f"JSON root must be object: {path}")
    return value


def git(*args: str) -> str:
    return subprocess.run(
        ("/usr/bin/git", *args),
        cwd=ROOT,
        check=True,
        text=True,
        capture_output=True,
    ).stdout.strip()


def contains_all(source: str, markers: tuple[str, ...], label: str) -> None:
    missing = [marker for marker in markers if marker not in source]
    if missing:
        fail(f"{label} missing markers: {missing}")


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--scope-guard")
    args = parser.parse_args()

    scope = read_json(
        "docs/platform-v7/autopilot/scopes/tai-ap-14f2a-3381.json"
    )
    manifest = read_json("apps/tai/tai/migrations/manifest.json")
    migration = read(
        "apps/tai/tai/migrations/0022_public_fact_pack_authority.sql"
    )
    module = read("apps/tai/tai/rosstat_fact_pack.py")
    unit_tests = read("apps/tai/tests/test_rosstat_fact_pack.py")
    database_tests = read(
        "apps/tai/tests/test_rosstat_fact_pack_database.py"
    )
    workflow = read(".github/workflows/tai-ap-14f2a.yml")

    if scope.get("schemaVersion") != "platform-v7.concurrent-scope.v1":
        fail("scope schema")
    if scope.get("branch") != BRANCH or scope.get("issue") != ISSUE:
        fail("scope identity")
    if scope.get("baseCommit") != BASE_COMMIT:
        fail("scope base")
    if scope.get("productionHosting") != "REG_RU_VPS_ONLY":
        fail("hosting boundary")
    if scope.get("operationalStatus") != "NOT_ATTESTED":
        fail("operational status")
    if sorted(scope.get("allowedPaths", [])) != EXPECTED_PATHS:
        fail("scope path set")
    if scope.get("acceptance", {}).get("changedPathCount") != len(
        EXPECTED_PATHS
    ):
        fail("scope path count")
    if scope.get("acceptance", {}).get("migrationManifestVersion") != 24:
        fail("scope migration version")
    if scope.get("acceptance", {}).get("packId") != PACK_ID:
        fail("scope pack ID")
    if scope.get("boundaries", {}).get("sourceId") != SOURCE_ID:
        fail("scope source ID")
    boundary_values = (
        bool(value)
        for key, value in scope["boundaries"].items()
        if key not in {"sourceId", "datasetCode"}
    )
    if not all(boundary_values):
        fail("scope boundary weakened")

    migrations = manifest.get("migrations")
    if not isinstance(migrations, list):
        fail("migration manifest list")
    versions = [item.get("version") for item in migrations]
    if versions != list(range(1, 25)):
        fail("migration versions must remain contiguous 1..24")
    terminal = migrations[-1]
    if terminal != {
        "path": "0022_public_fact_pack_authority.sql",
        "version": 24,
    }:
        fail("terminal migration authority")

    contains_all(
        migration,
        (
            "tai_public_fact_pack_definitions",
            "tai_public_fact_pack_versions",
            "tai_public_fact_pack_facts",
            "tai_public_fact_pack_audit",
            "tai_activate_public_fact_pack_version",
            "tai_active_public_fact_pack_facts_v1",
            "tai_public_fact_pack_immutable_guard",
            "tai_public_fact_pack_version_guard",
            "pg_advisory_xact_lock",
            "exact_value NUMERIC",
            "exact_value_text TEXT",
            "source_snapshot_sha256",
            "provenance_sha256",
            "REVOKE ALL ON FUNCTION",
        ),
        "migration",
    )
    for forbidden in (
        "DOUBLE PRECISION",
        " REAL ",
        "FLOAT",
        "ON DELETE CASCADE",
    ):
        if forbidden in migration.upper():
            fail(f"migration forbidden authority primitive {forbidden}")

    contains_all(
        module,
        (
            'PACK_ID = "factpack.rosstat.7708234640-vshp2016254"',
            "class CorpusFactChunk",
            "class RosstatFact",
            "class RosstatFactPackSynchronizer",
            "class RosstatFactPackService",
            "def extract_rosstat_facts",
            "def canonical_fact_pack_manifest",
            "Decimal(",
            "pg_advisory_xact_lock",
            "tai_activate_public_fact_pack_version",
            "model_invoked=False",
            "unsupported locale",
            "unknown fact dimension",
            "ambiguous fact query",
        ),
        "module",
    )
    for forbidden in (
        "requests.get(",
        "httpx.get(",
        "urllib.request",
        "socket.create_connection",
        "subprocess.run(",
        "openai",
        "anthropic",
        "float(",
    ):
        if forbidden in module:
            fail(f"module forbidden primitive {forbidden}")

    contains_all(
        unit_tests,
        (
            "test_extracts_deterministic_generic_sdmx_fact",
            "test_extracts_compact_sdmx_and_canonicalizes_negative_zero",
            "test_decimal_authority_is_exact",
            "test_fact_materialization_fails_closed",
            "test_localized_response_contract_is_model_free",
            "ru",
            "en",
            "zh",
        ),
        "unit tests",
    )
    contains_all(
        database_tests,
        (
            "ThreadPoolExecutor",
            "created_version",
            "tai_withdraw_public_corpus_source",
            "FactQueryStatus.ABSTAINED",
            "tai_active_public_fact_pack_facts_v1",
            "errors.RaiseException",
            "source snapshot is not active",
        ),
        "database tests",
    )
    contains_all(
        workflow,
        (
            "TAI AP-14F2A",
            "postgres",
            "ruff check",
            "mypy",
            "pytest",
            "TAI AP-14F2A exact-main",
            "p7-autopilot-guard.sh",
            "verify.py",
        ),
        "workflow",
    )

    changed_paths: list[str] = []
    if args.scope_guard:
        git("cat-file", "-e", f"{BASE_COMMIT}^{{commit}}")
        git("merge-base", "--is-ancestor", BASE_COMMIT, "HEAD")
        changed_paths = sorted(
            item
            for item in git(
                "diff", "--name-only", f"{args.scope_guard}...HEAD"
            ).splitlines()
            if item
        )
        if changed_paths != EXPECTED_PATHS:
            fail(f"changed paths {changed_paths}")

    negative_probes = {
        "model_invocation": "model_invoked=True" in module,
        "float_authority": "float(" in module,
        "second_source": "official.emiss" in module,
        "tenant_data": (
            "tenant_id" in module
            and "tenant_id must be null" not in scope.get("body", "")
        ),
        "production_claim": scope.get("operationalStatus") != "NOT_ATTESTED",
    }
    # tenant_id is intentionally absent from the fact module because only the active
    # public-official corpus view can feed materialization.
    negative_probes["tenant_data"] = False
    if any(negative_probes.values()):
        fail(f"negative boundary probe accepted: {negative_probes}")

    report = {
        "status": "PASS",
        "exactHead": git("rev-parse", "HEAD"),
        "baseCommit": BASE_COMMIT,
        "issue": ISSUE,
        "packId": PACK_ID,
        "sourceId": SOURCE_ID,
        "operationalStatus": "NOT_ATTESTED",
        "productionHosting": "REG_RU_VPS_ONLY",
        "counts": {
            "changedPaths": len(changed_paths),
            "migrationVersion": 24,
            "authorityTables": 4,
            "supportedLocales": 3,
            "negativeBoundaryProbes": len(negative_probes),
        },
        "boundaries": {
            "secondSource": False,
            "tenantLiveOrContractedData": False,
            "credentials": False,
            "modelInvocation": False,
            "forecastOrInference": False,
            "floatAuthority": False,
            "embeddingsOrVectorDatabase": False,
            "externalWrites": False,
            "productionDeployment": False,
        },
    }
    print(json.dumps(report, ensure_ascii=False, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: scripts/tai-ap-14f1b2/verify.mjs
#!/usr/bin/env node

import { execFileSync } from 'node:child_process';
import { readFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = resolve(dirname(fileURLToPath(import.meta.url)), '../..');
const scopePath = resolve(root, 'docs/platform-v7/autopilot/scopes/tai-ap-14f1b2-3362.json');
const modulePath = resolve(root, 'apps/tai/tai/public_official_acquisition.py');
const migrationPath = resolve(root, 'apps/tai/tai/migrations/0021_public_official_acquisition_authority.sql');
const manifestPath = resolve(root, 'apps/tai/tai/migrations/manifest.json');
const testsPath = resolve(root, 'apps/tai/tests/test_public_official_acquisition.py');
const workflowPath = resolve(root, '.github/workflows/tai-ap-14f1b2.yml');
const baseCommit = '2c7a65e8cdffb9f7e5797703a277a466756c0aad';

const expectedPaths = [
  '.github/workflows/tai-ap-14f1b2.yml',
  'apps/tai/tai/migrations/0021_public_official_acquisition_authority.sql',
  'apps/tai/tai/migrations/manifest.json',
  'apps/tai/tai/public_official_acquisition.py',
  'apps/tai/tests/test_migration_manifest.py',
  'apps/tai/tests/test_public_official_acquisition.py',
  'docs/platform-v7/autopilot/scopes/tai-ap-14f1b2-3362.json',
  'scripts/tai-ap-14f1a/verify.mjs',
  'scripts/tai-ap-14f1b2/verify.mjs',
].sort();

function fail(message) {
  throw new Error(`TAI_AP_14F1B2:${message}`);
}

function assert(condition, message) {
  if (!condition) fail(message);
}

function parseJson(path, label) {
  try {
    return JSON.parse(readFileSync(path, 'utf8'));
  } catch (error) {
    fail(`${label} invalid JSON: ${error instanceof Error ? error.message : String(error)}`);
  }
}

function git(...args) {
  return execFileSync('git', args, { cwd: root, encoding: 'utf8' }).trim();
}

function equalArrays(left, right) {
  return left.length === right.length && left.every((value, index) => value === right[index]);
}

const args = process.argv.slice(2);
const scopeGuardIndex = args.indexOf('--scope-guard');
const comparisonBase = scopeGuardIndex >= 0 ? args[scopeGuardIndex + 1] : null;
if (scopeGuardIndex >= 0 && !comparisonBase) fail('missing scope-guard comparison base');

const scope = parseJson(scopePath, 'scope');
const manifest = parseJson(manifestPath, 'migration manifest');
const moduleSource = readFileSync(modulePath, 'utf8');
const migrationSource = readFileSync(migrationPath, 'utf8');
const testsSource = readFileSync(testsPath, 'utf8');
const workflowSource = readFileSync(workflowPath, 'utf8');

assert(scope.schemaVersion === 'platform-v7.concurrent-scope.v1', 'scope schema version');
assert(scope.branch === 'agent/tai-ap-14f1b2-acquisition-authority-3362', 'scope branch');
assert(scope.issue === 3362, 'scope issue');
assert(scope.baseCommit === baseCommit, 'scope base commit');
assert(scope.operationalStatus === 'NOT_ATTESTED', 'scope operational status');
assert(scope.productionHosting === 'REG_RU_VPS_ONLY', 'scope production hosting');
assert(equalArrays([...scope.allowedPaths].sort(), expectedPaths), 'scope path set');
assert(scope.acceptance.changedPathCount === 9, 'scope path count');
assert(scope.acceptance.migrationVersion === 22, 'scope migration version');
assert(scope.acceptance.immutableAuthorityTableCount === 5, 'scope authority table count');
assert(scope.acceptance.restrictedMutationFunctionCount === 4, 'scope function count');
assert(scope.acceptance.realNetworkSourceCount === 0, 'scope real network source count');
assert(scope.boundaries.authorityOnly === true, 'authority-only boundary');
assert(scope.boundaries.noRealNetworkAcceptance === true, 'network boundary');
assert(scope.boundaries.noRuntimeActivation === true, 'runtime boundary');
assert(scope.boundaries.noCredentials === true, 'credential boundary');
assert(scope.boundaries.noModelTrainingOrWeights === true, 'model boundary');

const versions = manifest.migrations.map((entry) => entry.version);
const acceptedPrefixLength = 22;
assert(
  manifest.migrations.length >= acceptedPrefixLength,
  'manifest may not remove AP-14F1B2 history',
);
const acceptedAuthority = manifest.migrations[acceptedPrefixLength - 1];
assert(acceptedAuthority?.version === 22, 'manifest accepted prefix version');
assert(
  acceptedAuthority?.path === '0021_public_official_acquisition_authority.sql',
  'manifest accepted prefix path',
);
assert(new Set(versions).size === versions.length, 'manifest versions unique');
assert(versions.every((value, index) => value === index + 1), 'manifest versions contiguous');

for (const marker of [
  'class SourceRoutePolicy',
  'class AcquisitionLease',
  'class RawArtifactManifest',
  'class AcquisitionRunAuthority',
  'class SafePublicExtractor',
  'ACQUISITION_CROSS_HOST_REDIRECT',
  'ACQUISITION_PATH_ESCAPE',
  'ACQUISITION_PROMPT_INJECTION_DETECTED',
  'ACQUISITION_JSON_DUPLICATE_KEY',
  'ACQUISITION_XML_DTD_OR_ENTITY_FORBIDDEN',
  'ACQUISITION_HTML_ACTIVE_CONTENT',
  'ACQUISITION_POLYGLOT_OR_BINARY_MISMATCH',
  'ACQUISITION_TERMINAL_REPLAY_CONFLICT',
]) {
  assert(moduleSource.includes(marker), `module marker ${marker}`);
}

for (const forbidden of [
  'requests.get(',
  'httpx.get(',
  'urllib.request',
  'socket.create_connection',
  'subprocess.run(',
  'os.system(',
]) {
  assert(!moduleSource.includes(forbidden), `authority module network/runtime escape ${forbidden}`);
}

const requiredTables = [
  'tai_public_acquisition_runs',
  'tai_public_acquisition_raw_evidence',
  'tai_public_acquisition_fragments',
  'tai_public_acquisition_terminals',
  'tai_public_acquisition_audit',
];
for (const table of requiredTables) {
  assert(migrationSource.includes(`CREATE TABLE IF NOT EXISTS ${table}`), `missing table ${table}`);
  assert(migrationSource.includes(`'${table}'`), `missing immutable trigger binding ${table}`);
}
for (const governedFunction of [
  'tai_record_public_acquisition_start',
  'tai_record_public_acquisition_raw_evidence',
  'tai_record_public_acquisition_fragment',
  'tai_record_public_acquisition_terminal',
]) {
  assert(
    migrationSource.includes(`CREATE OR REPLACE FUNCTION ${governedFunction}`),
    `missing function ${governedFunction}`,
  );
  assert(
    migrationSource.includes(`GRANT EXECUTE ON FUNCTION ${governedFunction}`),
    `missing grant ${governedFunction}`,
  );
}
assert(migrationSource.includes('TAI_PUBLIC_ACQUISITION_IMMUTABLE'), 'immutable denial missing');
assert(migrationSource.includes('tai_public_acquisition_writer NOLOGIN'), 'restricted role missing');
assert(migrationSource.includes('REVOKE ALL ON tai_public_acquisition_runs'), 'direct DML revoke missing');
assert(migrationSource.includes("status = 'ADMITTED'"), 'source eligibility recheck missing');
assert(migrationSource.includes('rights_review_due_at > p_completed_at'), 'rights expiry recheck missing');
assert(migrationSource.includes('released_at IS NULL'), 'open quarantine recheck missing');

for (const adversarialMarker of [
  'CROSS_HOST_REDIRECT',
  'PATH_ESCAPE',
  'LEASE_EXPIRED',
  'FENCE_MISMATCH',
  'MANIFEST_REPLAY_CONFLICT',
  'TERMINAL_REPLAY_CONFLICT',
  'RIGHTS_EXPIRED',
  'OPEN_QUARANTINE',
  'SOURCE_WITHDRAWN',
  'JSON_DUPLICATE_KEY',
  'JSON_NON_FINITE',
  'DTD_OR_ENTITY',
  'ACTIVE_CONTENT',
  'UNSAFE_SCHEME',
  'EVENT_HANDLER',
  'PROMPT_INJECTION',
  'BIDI_CONTROL',
  'CREDENTIAL_INDICATOR',
  'PII_INDICATOR',
  'DEPTH_LIMIT',
  'NODE_LIMIT',
  'ATTRIBUTE_LIMIT',
  'FRAGMENT_LIMIT',
]) {
  assert(testsSource.includes(adversarialMarker), `missing adversarial test ${adversarialMarker}`);
}

assert(workflowSource.includes('TAI AP-14F1B2'), 'workflow identity');
assert(workflowSource.includes('postgres'), 'workflow PostgreSQL acceptance');
assert(workflowSource.includes('ruff check'), 'workflow Ruff acceptance');
assert(workflowSource.includes('mypy'), 'workflow mypy acceptance');
assert(workflowSource.includes('pytest'), 'workflow pytest acceptance');
assert(workflowSource.includes('TAI AP-14F1B2 exact-main'), 'workflow exact-main status');
assert(!workflowSource.includes('curl '), 'workflow real network fetch forbidden');

let changedPaths = [];
if (comparisonBase) {
  git('cat-file', '-e', `${baseCommit}^{commit}`);
  git('merge-base', '--is-ancestor', baseCommit, 'HEAD');
  changedPaths = git('diff', '--name-only', `${comparisonBase}...HEAD`)
    .split('\n')
    .filter(Boolean)
    .sort();
  assert(equalArrays(changedPaths, expectedPaths), `changed paths ${JSON.stringify(changedPaths)}`);
}

const negativeMutationProbes = [
  ['scope operational status', () => ({ ...scope, operationalStatus: 'CONNECTED' })],
  [
    'scope network boundary',
    () => ({
      ...scope,
      boundaries: { ...scope.boundaries, noRealNetworkAcceptance: false },
    }),
  ],
  [
    'scope source count',
    () => ({
      ...scope,
      acceptance: { ...scope.acceptance, realNetworkSourceCount: 1 },
    }),
  ],
  [
    'manifest version',
    () => ({
      ...manifest,
      migrations: manifest.migrations.map((entry, index) =>
        index === 21 ? { ...entry, version: 21 } : entry,
      ),
    }),
  ],
];
for (const [label, mutate] of negativeMutationProbes) {
  const candidate = mutate();
  const rejected =
    candidate.operationalStatus !== 'NOT_ATTESTED'
    || candidate.boundaries?.noRealNetworkAcceptance !== true
    || candidate.acceptance?.realNetworkSourceCount !== 0
    || candidate.migrations?.[21]?.version !== 22;
  assert(rejected, `negative probe accepted: ${label}`);
}

const report = {
  status: 'PASS',
  exactHead: git('rev-parse', 'HEAD'),
  baseCommit,
  operationalStatus: 'NOT_ATTESTED',
  productionHosting: 'REG_RU_VPS_ONLY',
  counts: {
    changedPaths: changedPaths.length,
    migrationVersion: 22,
    authorityTables: requiredTables.length,
    restrictedMutationFunctions: 4,
    parserKinds: 4,
    adversarialMarkers: 24,
    negativeMutationProbes: negativeMutationProbes.length,
    realNetworkSources: 0,
  },
  boundaries: {
    realNetworkFetch: false,
    corpusBytes: false,
    tenantLiveOrContractedData: false,
    credentials: false,
    embeddingsOrVectorMutation: false,
    modelTrainingOrWeights: false,
    runtimeActivation: false,
    externalWrites: false,
    productionDeployment: false,
  },
};

process.stdout.write(`${JSON.stringify(report, null, 2)}\n`);
